File: gclib/src/main/java/com/github/guqt178/utils/log/LoggerWrapper.java
package com.github.guqt178.utils.log;

import org.jetbrains.annotations.NotNull;

final class LoggerWrapper implements ILogger {

  private static volatile ILogger mLogger;

  private LoggerWrapper() {
    throw new IllegalStateException("Don`t instantiate ME!!");
  }

  public static ILogger getLogger() {
    if (null == mLogger) {
      synchronized (LoggerWrapper.class) {
        if (null == mLogger) {
          mLogger = new LoggerImpl();
        }
      }
    }
    return mLogger;
  }

  @Override
  public void wtf(@NotNull String msg, @NotNull Object... format) {
    getLogger().wtf(msg, format);
  }

  @Override
  public void warn(@NotNull String msg, @NotNull Object... format) {
    getLogger().warn(msg, format);
  }

  @Override
  public void info(@NotNull String msg, @NotNull Object... format) {
    getLogger().info(msg, format);
  }

  @Override
  public void error(@NotNull String msg, @NotNull Object... format) {
    getLogger().error(msg, format);
  }

  @Override
  public void debug(@NotNull String msg, @NotNull Object... format) {
    getLogger().debug(msg, format);
  }

  @Override
  public void tag(@NotNull String tag) {
    getLogger().tag(tag);
  }

  @NotNull
  @Override
  public String getTag() {
    return getLogger().getTag();
  }
}

File: modules/gclib/src/main/kotlin/com/github/guqt178/utils/log/LoggerWrapper.kt
package com.github.guqt178.utils.log

internal class LoggerWrapper private constructor() : ILogger {
    override fun wtf(msg: String, vararg format: Any) {
        logger!!.wtf(msg, *format)
    }

    override fun warn(msg: String, vararg format: Any) {
        logger!!.warn(msg, *format)
    }

    override fun info(msg: String, vararg format: Any) {
        logger!!.info(msg, *format)
    }

    override fun error(msg: String, vararg format: Any) {
        logger!!.error(msg, *format)
    }

    override fun debug(msg: String, vararg format: Any) {
        logger!!.debug(msg, *format)
    }

    override fun tag(tag: String) {
        logger!!.tag(tag)
    }

    override fun getTag(): String {
        return logger!!.getTag()
    }

    companion object {
        @Volatile
        private var mLogger: ILogger? = null

        @JvmStatic
        val logger: ILogger?
            get() {
                if (null == mLogger) {
                    synchronized(LoggerWrapper::class.java) {
                        if (null == mLogger) {
                            mLogger = LoggerImpl()
                        }
                    }
                }
                return mLogger
            }
    }

    init {
        throw IllegalStateException("Don`t instantiate ME!!")
    }
}